app/src/main/kotlin/com/coredocker/android/data/network/graphql/UsersApi.kt

```diff
@@ -0,0 +1,97 @@
+package com.coredocker.android.data.network.graphql
+
+import android.util.Log
+import com.apollographql.apollo.ApolloClient
+import com.apollographql.apollo.api.Input
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.coroutines.toDeferred
+import com.apollographql.apollo.coroutines.toFlow
+import com.coredocker.*
+import com.coredocker.fragment.CommandResultFragment
+import com.coredocker.fragment.RealTimeNotificationsMessageFragment
+import com.coredocker.fragment.UserFragment
+import com.coredocker.type.UserCreateUpdateModelInput
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.map
+import java.math.BigInteger
+
+
+class UsersApi(
+    private val _apolloClient: ApolloClient
+) {
+    private val _tag: String = UsersApi::class.java.simpleName
+
+    suspend fun queryAll(): PagedFragment<UserFragment> {
+
+        val response =  _apolloClient.query(AllUsersQuery()).toDeferred().await()
+        validateResponse(response, response.data?.users?.paged != null)
+        return PagedFragment(
+            response.data!!.users.paged.items.map { it -> it!!.fragments.userFragment },
+            BigInteger(response.data!!.users.paged.count.toString())
+        )
+    }
+
+    suspend fun addUser(
+        name: String,
+        email: String,
+        password: String?,
+        roles: List<String> = listOf("Guest")
+    ): CommandResultFragment {
```

Comment:
It would make sense to switch threads here from [CallingThread] to the IO thread, by changing this line to:

`): CommandResultFragment = withContext(Dispatchers.IO) {`

This means that everything inside this block will run on the IO thread, then return it on the thread it was called from. This means the calling class can run the function on whatever thread makes most sense for it to use.

The way you're doing it now, you're pushing the responsibility of determining the Thread selection to the calling class. See my other comment about GlobalScope.launch.

Using `withContext(dispatcher)` also means you can inject the context if you so wish, for testing purposes.

Reply:
AWESOME ... that makes sense. I will do it.